test/test_torch.py

```diff
@@ -2466,39 +2466,156 @@ def _test_rpow(self, cast):
         if not IS_WINDOWS:
-            typecasts.append(lambda x: x.int())
+            _test_integral_pow(torch.int32, (-10, 10), device)
 
-        shape = (11, 5)
-        tensor = cast(torch.LongTensor(shape).random_(-10, 10))
-        exps = [0, 1, 2, 5, cast(torch.LongTensor(shape).random_(0, 20))]
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_int_tensor_pow_neg_ints(self, device):
+        ints = [torch.iinfo(torch.int32).min,
+                -3, -2, -1, 0, 1, 2, 3,
+                torch.iinfo(torch.int32).max]
+        neg_ints = [torch.iinfo(torch.int32).min, -3, -2, -1]
+
+        tensor = torch.tensor(ints, dtype=torch.int32, device=device)
+        nparr = np.array(ints, dtype=np.int32)
+        out = torch.empty_like(tensor, device=device)
+
+        test_cases = [
+            lambda: tensor.pow(pow),
+            lambda: tensor.pow_(pow),
+            lambda: torch.pow(tensor, pow),
+            lambda: torch.pow(tensor, pow, out=out)
+        ]
 
-        for typecast in typecasts:
-            for exp in exps:
-                t = typecast(tensor)
-                e = exp if isinstance(exp, int) else typecast(exp)
-                check_against_np(t, e)
+        for pow in neg_ints:
+            self.assertRaisesRegex(
+                ValueError,
+                "Integers to negative integer powers are not allowed.",
+                lambda: np.power(nparr, pow))
+            for test_case in test_cases:
+                self.assertRaisesRegex(
+                    RuntimeError,
+                    "Integers to negative integer powers are not allowed.",
+                    test_case)
+
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_long_tensor_pow_floats(self, device):
+        ints = [0, 1, 23, 4567]
+        floats = [0.0, 1 / 3, 1 / 2, 1.0, 3 / 2, 2.0]
+
+        tensor = torch.tensor(ints, dtype=torch.int64, device=device)
+        nparr = np.array(ints, dtype=np.int64)
+
+        for pow in floats:
+            if device == 'cuda' and not TEST_WITH_ROCM:
+                # Current pow CUDA implementation casts exponent
+                # to tensor dtype, but numpy does not, that's why:
+                # pow CUDA  4 ^ 0.5 = 1
+                # numpy pow 4 ^ 0.5 = 2
+                # This line must be deleted as soon as
+                # pow CUDA implementation is fixed.
+                pow = int(pow)
+
+            # cpu CUDA implementation is already fixed and
+            # does not cast exponent to tensor dtype,
+            # that why it is compatible with numpy.
+            expected = np.power(nparr, pow).astype(np.int64)
+
+            actual = tensor.pow(pow)
+            self.assertEqual(expected, actual.cpu().numpy())
+
+            actual = tensor.clone()
+            actual2 = actual.pow_(pow)
+            self.assertEqual(expected, actual.cpu().numpy())
```

Comment:
We shouldn't be using self.assertEqual with numpy arrays. Also, usually the ordering is `self.assertEqual(result_value, expected_value)`.

Reply:
addressed